### PR TITLE
Add enemy info

### PR DESCRIPTION
**Pull request by gakada**
Includes enemy stats, equipment and stage1/stage2 aircraft counts for air power estimation.

Another possible extension for this schema is to use friendly air power calculated in the viewer, real air status and `$min` to record lower air power bounds for air statuses.

### CODE REVIEW

**Comment by KagamiChan** [comment]
what is the purpose for stage1 and stage2 plane remaining and loss without fleet air power?
seems we could sum up these to have a plane total, otherwise existing records will hardly get updated

**Comment by gakada** [comment]
stage1 count is total number of fighters and bombers in the enemy fleet (for api_kouku at least, LBAS can include recons as well), stage2 count is total number of bombers (after stage1 loss, stage1 loss include both fighters and bombers, but since stage1 loss is minimized and stage2 total is maximized in the limit it is just total number of bombers in the enemy fleet, also stage1 - stage2 is then total number of fighters). Extracting fighter/bomber counts for already known enemies it is possible to find fighter/bomber counts for unknown enemies, with equipment also known, it can be possible to estimate or to find exact air power for unknown enemies (as sum(floor(sqrt(slot) * aa)), with exact slot, or sometimes estimated, like setting 0 for non plane slots, or dividing between all bombers/fighters w.r.t. stage1/stage2 counts, or using numbers of some related enemy form, etc.).

Frequently new air power is estimated using this method (that is, not based on friendly air power tests), e.g., some [@Croshadow_](https://twitter.com/Croshadow_), [@syoukuretin](https://twitter.com/syoukuretin) tweets, https://nga.178.com/read.php?tid=13930924, https://nga.178.com/read.php?tid=13930924&pid=295296601 for current event.

**Comment by gakada** [comment]
As an example https://twitter.com/Croshadow_/status/1041280167832776704

dnneyegvsae66jz

Here is stats, equipment, stage1 api_e_count = 168, api_e_lostcount = 0, stage2 api_e_count = 110 (in general api_e_lostcount != 0 case can still be useful, maybe having all possible configurations is even more useful, but I used `$min`/`$max` to have less records to `update`), or extracting known slots of 1579, 168 - 96 = 72 total, 168 - 110 - 32 = 26 fighters, 72 - 26 = 46 bombers for unknown 1776, assuming even distribution (46 / 2) it is [26, 23, 23] for slots, using equipment stats, floor(sqrt(26) * 10) + floor(sqrt(23) * 8) + floor(sqrt(23) * 8) = 126 air power (111 in extreme distribution case like floor(sqrt(26) * 10) + floor(sqrt(1) * 8) + floor(sqrt(45) * 8)). Without knowing api_e_lostcount and stage2 api_e_count still can estimate 126 as floor(sqrt(24) * 10) + floor(sqrt(24) * 8) + floor(sqrt(24) * 8) (72 / 3) just from stage1 api_e_count. For enemies with one plane slot this can be more exact, but the point is stage1 loss and stage2 count can be useful.

**Comment by KagamiChan** [comment]
Thanks for the explanation, and sorry for being retarded. I think I somehow catch your idea.

one thing to confirm:
according to [the documentation](https://docs.mongodb.com/manual/reference/operator/update/), `$min` and `$max` only updates after comparing incoming value and existing value, is there any chance that a minimal `planeLost` and a maximal `planes2` do not come from the same battle (say, Record 1 updates `planeLost` and Record2 updates `planes2`)? And if so, will it lead to an inaccurate calculation?

By the way, would you like to detail the model description in a similar way as current documentation in wiki page?

**Comment by gakada** [comment]
Yeah, it can come from different battles, if stage 1 count is `f + b`, transitioning to stage 2 `(f - df) + (b - db)` with `df + db` loss and `b - db` count, then rather than storing values for individual battles, `$min` and `$max` are finding independent minimum and maximum for `df + db` and `b - db` for a given configuration `f + b`, most interesting case with loss = 0 will have the most information, always from the same battle.

Though based on this I think maybe it is more interesting to store `$max` stage 2 count (`b - db`) and `$min` stage 2 count + loss (`b + df`) instead. Loss = 0 case will have them equal (from the same battle), otherwise it is giving range `b - db <= b <= b + df` for bombers (from different battles).

Will do for wiki page. There is also some existing models that should be added.

**Comment by gakada** [comment]
Changed schema to planes/bombersMin/bombersMax.

File: controllers/api/v2/report.es [hunk 1]
```diff
@@ -270,6 +271,66 @@ router.post('/api/report/v2/ship_stat', async (ctx, next) => {
...
+      stats1,
+      equips1,
+      ships2,
+      levels2,
+      hp2,
+      stats2,
+      equips2,
+      planes,
+      // bombersMin,
+      // bombersMax,
```
Comment (KagamiChan) [inline controllers/api/v2/report.es]: please remove comments here

**Review by KagamiChan — CHANGES_REQUESTED** [review]
Sorry for keeping you waiting. I chose to put off this PR because of the fall event, and failed mind myself about it. All looks good except the commented lines

File: controllers/api/v2/report.es (hunk 1 above)
Comment (gakada) [reply controllers/api/v2/report.es]: done :ok_hand:

**Comment by KagamiChan** [comment]
LGTM cc @KochiyaOcean

**Comment by KagamiChan** [comment]
cc @yukixz

**Comment by gakada** [comment]
@KagamiChan I forgot to ask, is there a dump available for this collection?

**Comment by KagamiChan** [comment]
@gakada I've added a cron task to dump the data, I'll inform you once it's ready

**Comment by KagamiChan** [comment]
Sorry for being late

It will be available in 8 hours at https://poi.0u0.moe/dump/enemy-info.json

Currently you might also find the data named `enemy-stat.json`, it was my mistake

**Comment by gakada** [comment]
Thanks.

By the way, both `enemy-info.json` and `ship-stat.json` are not real JSON files, but newline separated lists of JSON objects, maybe it can be wrapped in a JSON array for easier use? If in bash, something like

```sh
to_json () {
  printf [ > $1.json
  tr '\n' ',' < $1 | head -c -1 >> $1.json
  printf ] >> $1.json
  # rm $1
}

# ship-stat and enemy-info as they are now
to_json ship-stat
to_json enemy-info
```

So that `require`/`readJson` can be used at once, instead of writing `readFile`/`split`/`map`/`JSON.parse`/etc.

**Comment by KagamiChan** [comment]
Thanks for sharing the script

It's true they're not valid json files, however it might be better to convert them at user side.

With such format, a processor does not have to load the whole file into memory, instead the data could be loaded and processed by lines. This is meaningful for importing into another mongo database, and also for scripts running on low end machines, if someone likes to
